### ISSUE TITLE
[Interest invokers] Keyboard inputs

### ISSUE THREAD

**Issue by mfreed7** [openui/open-ui#1133]
This issue mirrors https://github.com/openui/open-ui/issues/1052, but to discuss the keyboard interaction pattern specifically, for `interesttarget`.

Pulling from the conversation on https://github.com/openui/open-ui/issues/1052, a few things seem clear to me:

- Activating the `interestarget` on **keyboard focus** seems like a bad idea. Users who are merely trying to tab their way through a document will get lots of "noise" with additional visual items showing up. Worse, if the activated hovercard contains interactive elements, those will be *next in tab order* and will re-route the user through each such hovercard.
- Some might say that adding a delay (i.e. focus an element *for a period of time*) alleviates the above problem, but it merely adds confusion. Users will now have to tab very quickly through documents for fear of accidentally activating hovercards along the way. The same is true of mouse users hovering elements; however, mouse users have the luxury of being able to move the mouse all the way out of the way, and then later back to the same spot within the document. Keyboard users have no such luxury - they are in a linked list of focusable elements, and can't easily traverse out of order.
- If focus cannot be the `interesttarget` trigger, then by definition the trigger must be another keyboard hotkey. The task becomes:
    1. Choosing a hotkey, and standardizing it, or...
    2. Allowing user agents to select the hotkey, and provide a way for users to discover the proper hotkey

I think the best for users would be <span>#</span>1, so that the same hotkey activates all hovercards, across browsers and platforms. Ideally this should be an easy hotkey to press, not requiring modifier keys, etc. Perhaps the best, most self-discoverable such hotkey would be either the **up or down arrow**?

My fear is that it'll be hard to find an available hotkey that works across all platforms and browsers. But perhaps the arrow keys might be one such option? I don't know what happens when you have something like a radio button group where each radio has an `interesttarget`, or similar situations. Perhaps we could just make all such conflicts illegal (ignore `interesttarget`)?

**Comment by lukewarlow** [comment]
I think if we had a hotkey we'd need to make sure it's remappable in browser settings per WCAG.

Arrow keys would be interesting, though idk if that would mess with navigation for stuff like focus groups where it's perhaps legit to have an interest target.

Coming up with a new pattern on top of a potential novel touch system does concern me from an is this thing going to be speccable perspective. Both technically and politically. But if it's necessary then it's obviously worth it.

**Comment by nmn** [comment]
> Activating the interestarget on keyboard focus seems like a bad idea. Users who are merely trying to tab their way through a document will get lots of "noise" with additional visual items showing up.

I want to push back on this a bit. I think the "hovercard" pattern means showing the `interesttarget` on focus. The behavior is also consistent with how the content appears on "hover" when considering mouse interactions. There is also a major discoverability problem to hide the interesttarget behind a special hotkey.

However, I do agree with this part:

> Worse, if the activated hovercard contains interactive elements, those will be next in tab order and will re-route the user through each such hovercard.

I think the way to solve this conflict is by having *two* states for the `interesttarget`:
1. An `inert` preview only
2. An "activated" state where it behaves more like a dialog and the interactive elements within become focusable.

We've used a similar pattern for Facebook recently and the pattern seems to be working very well. Here are some of the details:

- When a `HovercardTrigger` link is focused, the `Hovercard` becomes visible, but it is `inert` and all the content withing is `aria-hidden` and unfocusable.
  - **There is one exception**: There is a small helper text that says something like "press the up arrow key to access link shortcuts"
  - This helper text has an `id` and it is used as the `aria-hint` value on the `HovercardTrigger` link
- When the user presses the chosen hot-key, (up and down arrow keys in our case), the Hovercard becomes "active" and starts to behave like an interactive modal dialog.
  - The focus is moved to the first interactive element within the hovercard.
  - There is focus trapping
  - There is visible `Close` button
  - Pressing `escape` closes the hovercard and returns the focus to the `HovercardTrigger` link
- Pressing `escape` also always hides the `Hovercard` inert preview. Giving the ability to use the up and down arrow keys for scrolling again.

### Extra Notes

- I do see the possible concern that *showing* a preview on focus can feel overwhelming and -busy, but I think the focus order problems can be solved by having the two "preview" and "active" modes for the interest target.
- The "visual treatment" for what appears on focus is up for debate. I could see an argument for letting the UserAgent choose the best treatment here.
  - It might make sense to show the entire hovercard (in an inert state) in some scenarios;
  - and, to show a much more subtle indicator in other scenarios.
- There are concerns today that taking over the `up` and `down` arrow keys is problematic because they are normally used for scrolling. Our rationale was that it is more discoverable for keyboard users that are used to Focus Group patterns. As a web standard and different hot key could be chosen which users would learn to use.
- There is a question about the "delay" after focusing a link similar to the delay after hovering a link.
- One big *benefit* of this approach is that it unifies the mouse and keyboard interactions and makes it easy to switch between the two input methods easily.
  - Hovering on the link shows the inert preview of the hovercard
  - Hovering on the hovercard "activates" it and moves focus to it behind the scenes.

**Comment by mfreed7** [comment]
Thanks for the points about keyboard focus. Let's Agenda+ to discuss it.

I've also added a bit more detail about the current direction to the explainer, in this PR: https://github.com/openui/open-ui/pull/1155. Of course, any/all of that can be changed if we need to change directions.

**Comment by mfreed7** [comment]
Awesome discussion today - thanks for the great feedback. Based on that conversation, a somewhat new direction seems to be emerging, along the lines of @nmn's [comment](https://github.com/openui/open-ui/issues/1133#issuecomment-2640866978). I've tried to summarize that new direction here - please feel free to correct me where I missed something, or if you have additional suggestions for improvement:

- trigger "interest" in the element when it receives keyboard focus, after `interest-target-show-delay` seconds.
- "lose interest" in the element when focus *leaves* the interest invoker or its target, after `interest-target-hide-delay` seconds.
- When activated via keyboard only (these additional behaviors do *not* apply when activated via mouse or touchscreen):
    - the target popover is shown, but is rendered with `interactivity: inert`. This keeps the popover from inserting any additional tab stops into the sequential focus navigation order.
    - a hotkey (e.g. ALT-UP) will cause the popover to be "fully activated" which removes the `interactivity` rule. This makes it non-inert and available in the focus navigation order. Potentially the first focusable element within the popover should also be focused at this point? Maybe not.
    - a new pseudo class (e.g. `:has-partial-interest` or similar) will match on the *target popover* only when it is in the "partially activated" inert state. It's possible this makes more sense (semantically) to match on the interest invoker, but it seems easier to use on the target.
    - a new UA stylesheet will will be added: `:has-partial-interest::after {content: "Press ALT-UP to activate this content"; interactivity: auto}` which adds (developer-overridable) hints to the user about the hotkey. Note the `interactivity` value here - that makes the hint text non-inert so it shows up in the a11y tree. Additionally it should likely get special treatment in the a11y tree so that it gets announced.

I believe this meets the following use case requirements:
 - Better user activation story for non-interactive "tooltips" that shouldn't be so hard to activate. I.e. just focus an element to see the tooltip.
 - Better discoverability story for rich "hovercards", because of the `::after` text.
 - Less risk of annoyance for keyboard users, since rich hovercards don't insert themselves automatically into the sequential focus order.
 - Ability to override inertness for key use cases, such as menus (where the author would add `:has-partial-interest {interactivity:auto}` to show immediately on focus) and "large/obtrusive" tooltips (where the author would add `:has-partial-interest {display:none}` so content doesn't show on focus).
 - The partial interest state can be indicated to the user (e.g. with `:has-partial-interest {opacity:50%}`) if desired, and the helpful text about the hotkey can be customized or hidden if desired.
 - The default state "just works" for keyboard users, in the case that the developer just tested/developed using a mouse, and didn't do any of the above.


A few side notes about the conversation:
- We talked about the part of @nmn's comment about the hovercard being a modal dialog. He pointed out that this was just to be able to build this functionality with existing tools (i.e. the lack of `interesttarget`), and isn't a desired outcome. So it's ok to maintain [our previous position](https://github.com/openui/open-ui/issues/1064#issuecomment-2596642781) that only popovers are supported by `interesttarget`, and not dialogs.
- @smhigley will open a separate issue about the potential need to be able to both cancel "interest" via preventDefault() and also directly trigger interest programmatically. This is to meet the needs of (among other things) zooming software, which might want to provide a more affirmative way to show interest that doesn't get in the way.
- The menu use case continues to feel a bit bolted on. We should continue to evaluate that use case as we go along and make sure it still makes sense, and is right to build with this API. I *think* it still makes sense, and the above is actually an improvement by a bit. But let's keep discussing.

**Comment by nmn** [comment]
> a new pseudo class (e.g. :has-partial-interest or similar) will match on the target popover only when it is in the "partially activated" inert state. It's possible this makes more sense (semantically) to match on the interest invoker, but it seems easier to use on the target.

Thinking of the use-cases I would want this on the target more than the invoker. The invoker can already use `:hover` and `:focus-visible` instead.

> Additionally it should likely get special treatment in the a11y tree so that it gets announced.

Adding context, but our current implementation uses `aria-describedby` which announces at a low-priority and various screen readers have the option to turn off such announcements. I think this is the right level of importance for this announcement as once the pattern is commonplace, users may not need it anymore.

> Ability to override inertness for key use cases, such as menus (where the author would add :has-partial-interest {interactivity:auto} to show immediately on focus) and "large/obtrusive" tooltips (where the author would add :has-partial-interest {display:none} so content doesn't show on focus).

These pseudo classes tie everything together SO WELL!!

> hovercard being a modal dialog... and isn't a desired outcome.

Just want to clarify that I'm more ambivalent on this issue. I missed the previous discussion on the reasons for choosing `popover` and `dialog`, but it might worth a reconsideration once the rest of the decisions are made to figure out the best solution for a11y.

> directly trigger interest programmatically. 

I assume this would only be for "full-interest" and not "partial-interest"? (I'll follow the other issue)

> The menu use case continues to feel a bit bolted on.

Many of the use-cases I can think of hover interactions that are purely menus would probably be solved by having a CSS way to access `top-layer`.

However, I can also think of many use-cases where interest-target is better suited for menu-like interactions. One that comes to mind is the alternate key options you see when you long press a key on a phone's virtual keyboard. It's a something that works with a simple click. But if you linger on it (with focus/hover/or press), it reveals some additional options.

Side-note: "linger" feels like a good word to describe all the "interest" interactions.

**Comment by nmn** [comment]
Great summary!

Small addition to the second point:

> We reiterated (mostly) that interesttarget should only invoke popovers (including <dialog popover>) but should not invoke modal dialogs. Mostly the reasoning is the https://github.com/openui/open-ui/issues/1064#issuecomment-2596642781, with the added point that with the new "partially activated" state, a modal dialog would cause the entire page to be non-interactive.

*If* we were to use modal dialogs, we would have to only use that role for "full-interest", and we would need an "inert" or some other role during "partial-interest".

We are pretty much in alignment that one way (`inert`) or the other (`focus-index="-1"`), the interest targets will not be part of the focus tree by default and will require some kind of explicit keyboard interaction to "activate".

In such a scenario, the question is mostly about the best interaction pattern for when the focus is *within* the interest target. I think depending on the use-case a focus trap may or may not be desirable. 
- For interest targets that behave like "addendums" having the focus move back out and to the next element in the tree makes sense.
- For interest targets that are more like "mini pages", a dialog-like focus trap might make more sense where you need to explicitly escape back *out*.

---

Also a question about `inert`.

Assuming that the interest target isn't going to be focusable anyway, what is the functional difference of using `inert` or not?

Would the entire content of the "partial-interest" interest target be used as the `aria-hint`? Would that work correctly when the interest target contains complex content?

**Comment by bramus** [comment]
> > Activating the interestarget on keyboard focus seems like a bad idea. Users who are merely trying to tab their way through a document will get lots of "noise" with additional visual items showing up.
> 
> I want to push back on this a bit. I think the "hovercard" pattern means showing the `interesttarget` on focus.

Found my way here to say the same thing after trying https://cdpn.io/pen/debug/raNLLor _(in Chrome Canary with flags)_. Good to see this was discussed and incorporated.

**Comment by mfreed7** [comment]
> > I want to push back on this a bit. I think the "hovercard" pattern means showing the `interesttarget` on focus.
> 
> Found my way here to say the same thing after trying https://cdpn.io/pen/debug/raNLLor _(in Chrome Canary with flags)_. Good to see this was discussed and incorporated.

Canary doesn't yet have this behavior (showing interest on focus), but [it will soon](https://chromium-review.googlesource.com/c/chromium/src/+/6418573).

**Comment by nmn** [comment]
just a quick heads up that before leaving, I rolled out the new hovercard implementation to 100% of public, if you want to play around with a user-space implementation to see how it feels.
